### DISCUSSION TITLE
Variable time partition start_date

### DISCUSSION THREAD

**Discussion by jPinhao-Rover** [dagster-io/dagster#19227] [Q&A] [answered]
I have a set of assets (dbt models) which I'm looking to implement partitions for. We group our assets using a "frequency" tag in the dbt model config, which defines whether the asset should be materialised hourly / daily / etc. Eg.:

```
@dbt_assets(
    manifest=dbt_manifest_path,
    select="tag:frequency:daily"
)
def dbt_daily_assets(context: AssetExecutionContext, dbt: DbtCliResource):
    dbt_run_invocation = dbt.cli(["run"], context=context)
    yield from dbt_run_invocation.stream()


daily_job = define_asset_job(
    name="dbt_daily",
    selection=[dbt_daily_assets]
)

daily_schedule = ScheduleDefinition(
            cron_schedule="12 0 * * *", job=daily_job
        )
```

Now as part of defining partitions, we need to define [the partition `start_date`](https://docs.dagster.io/_apidocs/partitions#dagster.DailyPartitionsDefinition), ie. the earliest partition for the asset.

These are added into the asset deinfition as part of the PartitionsDefinition, eg.:

```
@dbt_assets(
    manifest=dbt_manifest_path,
    select="tag:frequency:daily",
    partitions_def=DailyPartitionsDefinition(start_date={some_start_date}),
)
def dbt_daily_assets(context: AssetExecutionContext, dbt: DbtCliResource):
    dbt_run_invocation = dbt.cli(["run"], context=context)
    yield from dbt_run_invocation.stream()
``` 

The `start_date` can vary depending on the asset, and we'd like a way to configure our assets so they share the asset materialisation logic, partitioning, scheduling, etc. **BUT** with distinct `start_date`s. Ideally the `start_date` would be defined within the dbt model config and parsed in similarly to the `tag:frequency:daily`.

How would you approach defining these assets where everything is the same other than the parition start_date?

**Comment by OwenKephart** [comment, marked as answer]
Hi @jPinhao-Rover, unfortunately, all assets within a single `@dbt_assets` must have a single `partitions_def`, and two partitions definitions with a different `start_date` are considered distinct.

So you would need a separate decorated function for each combination of (frequency, start_date). This could in theory be done programmatically, i.e. something along the lines of:

```python
def get_dbt_assets(manifest):
    
    all_assets = []
    for start_date in get_start_dates(manifest):
        @dbt_assets(name=..., partitions_def=...)
        def _dbt_assets(...): ...
     
        all_assets.append(_dbt_assets)
    return all_assets
```
